LLama/Extensions/LLamaExecutorExtensions.cs

```diff
@@ -0,0 +1,169 @@
+                };
+            }
+        }
+
+        /// <summary>Format the chat messages into a string prompt.</summary>
+        private string CreatePrompt(IList<ChatMessage> messages)
+        {
+            if (messages is null)
+            {
+                throw new ArgumentNullException(nameof(messages));
+            }
+
+            ChatHistory history = new();
+
+            if (_executor is not StatefulExecutorBase seb ||
+                seb.GetStateData() is InteractiveExecutor.InteractiveExecutorState { IsPromptRun: true })
+            {
+                foreach (var message in messages)
+                {
+                    history.AddMessage(
+                        message.Role == ChatRole.System ? AuthorRole.System :
+                        message.Role == ChatRole.Assistant ? AuthorRole.Assistant :
+                        AuthorRole.User,
+                        string.Concat(message.Contents.OfType<TextContent>()));
+                }
+            }
+            else
+            {
+                // Stateless executor with IsPromptRun = false: use only the last message.
+                history.AddMessage(AuthorRole.User, string.Concat(messages.LastOrDefault()?.Contents.OfType<TextContent>() ?? []));
+            }
+         
+            return _historyTransform.HistoryToText(history);
+        }
+
+        /// <summary>Convert the chat options to inference parameters.</summary>
+        private static InferenceParams? CreateInferenceParams(ChatOptions? options)
+        {
+            List<string> antiPrompts = new(s_antiPrompts);
+            if (options?.AdditionalProperties?.TryGetValue(nameof(InferenceParams.AntiPrompts), out IReadOnlyList<string>? anti) is true)
+            {
+                antiPrompts.AddRange(anti);
+            }
+
+            return new()
+            {
+                AntiPrompts = antiPrompts,
+                TokensKeep = options?.AdditionalProperties?.TryGetValue(nameof(InferenceParams.TokensKeep), out int tk) is true ? tk : s_defaultParams.TokensKeep,
+                MaxTokens = options?.MaxOutputTokens ?? 256, // arbitrary upper limit
+                SamplingPipeline = new DefaultSamplingPipeline()
+                {
+                    AlphaFrequency = options?.AdditionalProperties?.TryGetValue(nameof(DefaultSamplingPipeline.AlphaFrequency), out float af) is true ? af : s_defaultPipeline.AlphaFrequency,
+                    AlphaPresence = options?.AdditionalProperties?.TryGetValue(nameof(DefaultSamplingPipeline.AlphaPresence), out float ap) is true ? ap : s_defaultPipeline.AlphaPresence,
+                    PenalizeEOS = options?.AdditionalProperties?.TryGetValue(nameof(DefaultSamplingPipeline.PenalizeEOS), out bool pe) is true ? pe : s_defaultPipeline.PenalizeEOS,
+                    PenalizeNewline = options?.AdditionalProperties?.TryGetValue(nameof(DefaultSamplingPipeline.PenalizeNewline), out bool pn) is true ? pn : s_defaultPipeline.PenalizeNewline,
```

Comment:
The spellchecker has a false positive on this line. If you add `pn` to the typos.toml file that should suppress it. (docs: https://github.com/marketplace/actions/typos-action#false-positives)

Reply:
I'll just rename it. Thanks.